stdlib/string.ml

```diff
@@ -280,3 +280,100 @@ let get_int32_le s i = B.get_int32_le (bos s) i
 let get_int32_be s i = B.get_int32_be (bos s) i
 let get_int64_le s i = B.get_int64_le (bos s) i
 let get_int64_be s i = B.get_int64_be (bos s) i
+
+(* Spellchecking *)
+
+let uchar_utf_8_byte_decode_length = function
+  | '\x00' .. '\x7F' -> 1
+  | '\x80' .. '\xC1' -> 0
+  | '\xC2' .. '\xDF' -> 2
+  | '\xE0' .. '\xEF' -> 3
+  | '\xF0' .. '\xF4' -> 4
+  | _ -> 0
+
+let utf_8_uchar_length s =
+  let slen = length s in
+  let i = ref 0 and ulen = ref 0 in
+  while (!i < slen) do
+    let dec_len = uchar_utf_8_byte_decode_length (unsafe_get s !i) in
+    i := (!i + if dec_len = 0 then 1 (* count one Uchar.rep *) else dec_len);
+    incr ulen;
+  done;
+  !ulen
+
+let uchar_array_of_utf_8_string s =
+  let slen = length s in (* is an upper bound on Uchar.t count *)
+  let uchars = Array.make slen Uchar.max in
+  let k = ref 0 and i = ref 0 in
+  while (!i < slen) do
+    let dec = get_utf_8_uchar s !i in
+    i := !i + Uchar.utf_decode_length dec;
+    uchars.(!k) <- Uchar.utf_decode_uchar dec;
+    incr k;
+  done;
+  uchars, !k
+
+let edit_distance' ?(limit = Int.max_int) s (s0, len0) s1 =
+  if limit <= 1 then (if equal s s1 then 0 else limit) else
+  let[@inline] minimum a b c = Int.min a (Int.min b c) in
+  let s1, len1 = uchar_array_of_utf_8_string s1 in
+  let limit = Int.min (Int.max len0 len1) limit in
+  if Int.abs (len1 - len0) >= limit then limit else
+  let s0, s1 = if len0 > len1 then s0, s1 else s1, s0 in
+  let len0, len1 = if len0 > len1 then len0, len1 else len1, len0 in
+  let rec loop row_minus2 row_minus1 row i len0 limit s0 s1 =
+    if i > len0 then row_minus1.(Array.length row_minus1 - 1) else
+    let len1 = Array.length row - 1 in
+    let row_min = ref Int.max_int in
+    row.(0) <- i;
+    for j = Int.max 1 (i - limit - 1) to Int.min len1 (i + limit + 1) do
+      let cost = if Uchar.equal s0.(i-1) s1.(j-1) then 0 else 1 in
+      let min = minimum
+          (row_minus1.(j-1) + cost) (* substitute *)
+          (row_minus1.(j) + 1)      (* delete *)
+          (row.(j-1) + 1)           (* insert *)
+      in
+      let min =
+        if (i > 1 && j > 1 &&
+            Uchar.equal s0.(i-1) s1.(j-2) &&
+            Uchar.equal s0.(i-2) s1.(j-1))
+        then Int.min min (row_minus2.(j-2) + cost) (* transpose *)
+        else min
+      in
+      row.(j) <- min;
+      row_min := Int.min !row_min min;
+    done;
+    if !row_min >= limit then (* can no longer decrease *) limit else
+    loop row_minus1 row row_minus2 (i + 1) len0 limit s0 s1
+  in
+  let ignore =
+    (* Value used to make the values around the diagonal stripe ignored
+       by the min computations when we have a limit. *)
+    limit + 1
```

Comment:
Using `limit` is sufficient, since this is already the maximal element of the implicit distance type.

Reply:
Indeed but it's easier to follow what happens when you debug with a "don't care value".